### PR TITLE
fix(framework:cli): Allow hypens in flwr new CLI

### PR DESCRIPTION
**Pull request by charlesbvll**
### Description

From [this discussion](https://github.com/adap/flower/pull/3256#discussion_r1573178415):

```
The directory name of the project should allow - (quite common in Python library names). The Python package inside will then use _ instead of - to make it a valid Python package name.
Example:

Project name: flwr-datasets (e.g., pip install flwr-datasets)
Package name: flwr_datasets
```

### Related issues/PRs

#3256 

### Explanation

Allow project names with hyphens and replace hyphens with underscores when creating the package.

### Checklist

- [x] Implement proposed change
- [x] Update the changelog entry below
- [x] Make CI checks pass
- [x] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### Changelog entry

<skip>

### Any other comments?

N/A

### CODE REVIEW

File: src/py/flwr/cli/new/new.py [hunk 1]
```diff
@@ -124,27 +125,37 @@ def new(
 
     # Set project directory path
     cwd = os.getcwd()
-    pnl = project_name.lower()
-    project_dir = os.path.join(cwd, pnl)
+    module_name = re.sub(r"[-_.]+", "-", project_name).lower()
```
Comment (danieljanes) [inline src/py/flwr/cli/new/new.py:128]: Should we remove `.` here?
```suggestion
    distribution_name = re.sub(r"[-_]+", "-", project_name).lower()
```

Comment (danieljanes) [reply src/py/flwr/cli/new/new.py:128]: Or should it be:
```suggestion
    module_name = re.sub(r"[-_.]+", "_", project_name).lower()
```

File: src/py/flwr/cli/new/new_test.py [hunk 2]
```diff
@@ -35,7 +35,7 @@ def test_render_template() -> None:
     """Test if a string is correctly substituted."""
     # Prepare
     filename = "app/README.md.tpl"
-    data = {"project_name": "FedGPT"}
+    data = {"project_name": "FedGPT", "module_name": "fedgpt", "import_name": "fedgpt"}
```
Comment (danieljanes) [inline src/py/flwr/cli/new/new_test.py:38]: Let's add one test case for each of the conversions and error situations handled by `flwr new`

File: src/py/flwr/cli/new/new.py (hunk 1 above)
Comment (danieljanes) [reply src/py/flwr/cli/new/new.py:128]: Or should it be:
```suggestion
    package_name = re.sub(r"[-_.]+", "-", project_name).lower()
```

File: src/py/flwr/cli/new/templates/app/pyproject.sklearn.toml.tpl [hunk 3]
```diff
@@ -18,3 +18,7 @@ dependencies = [
 
 [tool.hatch.build.targets.wheel]
 packages = ["."]
+
+[flower.components]
+serverapp = "$module_name.server:app"
+clientapp = "$module_name.client:app"
```
Comment (danieljanes) [inline src/py/flwr/cli/new/templates/app/pyproject.sklearn.toml.tpl:24]:
```suggestion
serverapp = "$import_name.server:app"
clientapp = "$import_name.client:app"
```

File: src/py/flwr/cli/new/templates/app/pyproject.tensorflow.toml.tpl [hunk 4]
```diff
@@ -20,5 +20,5 @@ dependencies = [
 packages = ["."]
 
 [flower.components]
-serverapp = "$project_name.server:app"
-clientapp = "$project_name.client:app"
+serverapp = "$module_name.server:app"
+clientapp = "$module_name.client:app"
```
Comment (danieljanes) [inline src/py/flwr/cli/new/templates/app/pyproject.tensorflow.toml.tpl:24]:
```suggestion
serverapp = "$import_name.server:app"
clientapp = "$import_name.client:app"
```

File: src/py/flwr/cli/new/templates/app/pyproject.sklearn.toml.tpl (hunk 3 above)
Comment (charlesbvll) [reply src/py/flwr/cli/new/templates/app/pyproject.sklearn.toml.tpl:24]: @danieljanes In this case we actually want the module name, i.e, the same name as the folder containing the Python files, not the import name (with dashes replaced with underscores)

Comment (charlesbvll) [reply src/py/flwr/cli/new/templates/app/pyproject.sklearn.toml.tpl:24]: From tests I made, this corresponds to the `package name` (dots, underscores and capital letters allowed)

**Review by tanertopal — APPROVED** [review]

**Pull request merged**